include/motor.hpp

```diff
@@ -0,0 +1,242 @@
+#ifndef MOTOR_H
+#define MOTOR_H
+
+#include "Arduino.h"
+#include <exception.h>
+
+#include "freertos/FreeRTOS.h"
+#include "freertos/task.h"
+#include "esp_log.h"
+//#include "freertos/semphr.h"
+
+/**
+ *  Motor
+ *   
+ *  position = mm
+ *  speed = mm/s
+ *  acceleration = mm/s^2
+ */
+
+#define DEBUG_CLIPPING              // Show debug messages when motions violating the machine 
+                                    // physics are commanded
+
+typedef struct {
+  float start;
+  float end;
+  float keepout; /*  Soft endstop preventing hard crashes in mm. Will be 
+                  *  subtracted twice from physicalTravel. Should be 
+                  *  sufficiently to completley drive clear from 
+                  *  homing switch */
+} MachineGeometry;
+
+#define MOVEMENT_TASK_FLAG_EXIT (1 << 0)
+#define MOVEMENT_TASK_FLAG_NEXT (1 << 1)
+
+// TODO - Change flags into a FreeRTOS Event Group which allows better synchronizing on Flag State
+#define MOTOR_FLAG_ENABLED (1 << 0)
+
+#define MOTOR_FLAG_WARNING (1 << 1)
+#define MOTOR_FLAG_ERROR (1 << 2)
+#define MOTOR_FLAG_FATAL (1 << 3)
+
+#define MOTOR_FLAG_AT_TARGET (1 << 4)
+#define MOTOR_FLAG_MOTION_ACTIVE (1 << 5)
+#define MOTOR_FLAG_HOMED (1 << 6)
+
+#define MOTOR_FLAG_OVERHEATING (1 << 7)
+#define MOTOR_FLAG_OVERFORCE (1 << 8)
+
+typedef enum {
+  INACTIVE = 0,
+  ACTIVE = 1,
+  HOMING = 2, 
+  ERROR = 3
+} MotorState;
+
+typedef enum {
+  // Retraction Emergency useful for Throat play, where a quick retraction to a safe position is required
+  RETRACT = 0, 
+
+  // De-energize Emergency useful for Vaginal/Anal play, where a retraction could cause pain.
+  // Instead the motor itself is de-energized so no Holding Torque is experienced.
+  DEENERGIZE = 1 
+} MotorEmergency
+
+class MotorException extends std::exception {
+
+}
+
+class MotorInterface {
+  public:
+    void enable() { this->addStatusFlag(MOTOR_FLAG_ENABLED); }
+    // TODO - virtual void disable();
+
+    SemaphoreHandle_t takeSemaphore() {
```

Comment:
It's not finished, but I'm working on adding some locking functionality, so the Motor can be driven in two modes.
- Direct (Via using goToPos with no Motion Task)
- Task-based (Via a FreeRTOS task which delivers new Motion Commands)

Direct won't have locking to an exclusive task, but will still have Semaphores preventing multiple Motion Commands being run at once.

Task-based attached a Task to the Motor, and no other Task / Code is allowed to use the goToPos and other motion commands. It'll check the Task Handle to make sure the Task matches the one which took out the `taskSemaphore`

Reply:
I'm not sure if Direct mode is really needed

Reply:
I would like to have it there. It comes in handy when troubleshooting.